osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/SecretManagementService.java

```diff
@@ -1,181 +1,360 @@
-                    SecretType.E_METER_ENCRYPTION_KEY_UNICAST);
-            GetSecretsResponse response = secretManagementClient.getSecretsRequest(request);
-            Optional<TypedSecret> optionalTypedSecret = getTypedSecretFromSoapResponse(response,
-                    SecretType.E_METER_ENCRYPTION_KEY_UNICAST);
+    private void validateGetResponse(List<SecurityKeyType> keyTypes, GetSecretsResponse response) {
+        this.validateOsgpResultAndTypedSecrets(response.getResult(), response.getTechnicalFault(),
+                response.getTypedSecrets(), keyTypes.size());
+    }
 
-            byte[] decryptedKey = decryptSoapSecret(deviceIdentification, optionalTypedSecret.orElseThrow(
-                    () -> new IllegalStateException("Secret not found:" + deviceIdentification)));
+    private void validateGetNewResponse(List<SecurityKeyType> keyTypes, GetNewSecretsResponse response) {
+        this.validateOsgpResultAndTypedSecrets(response.getResult(), response.getTechnicalFault(),
+                response.getTypedSecrets(), keyTypes.size());
+    }
 
-            log.trace("DlmsGlobalUnicastEncryptionKey for device " + deviceIdentification + " is " + Hex.encodeHexString(decryptedKey));
+    private Map<SecurityKeyType, byte[]> convertSoapSecretsToSecretMapByType(List<TypedSecret> soapSecrets) {
+        Function<TypedSecret, SecurityKeyType> convertType = ts -> SecurityKeyType.fromSecretType(ts.getType());
+        Function<TypedSecret, byte[]> convertSecret = ts -> this.decryptSoapSecret(ts);
+        return soapSecrets.stream().collect(Collectors.toMap(convertType, convertSecret));
+    }
 
-            return decryptedKey;
+    private GetSecretsRequest createGetSecretsRequest(String deviceIdentification, List<SecurityKeyType> keyTypes) {
+        GetSecretsRequest request = new GetSecretsRequest();
+        request.setDeviceId(deviceIdentification);
+        request.setSecretTypes(new SecretTypes());
+        List<SecretType> secretTypeList = request.getSecretTypes().getSecretType();
+        keyTypes.stream().forEach(kt -> secretTypeList.add(kt.toSecretType()));
+        return request;
+    }
 
-        } catch (Exception e) {
-            LOGGER.error("Error while retrieving encryption key", e);
-        }
-        return new byte[0];
+    private GetNewSecretsRequest createGetNewSecretsRequest(String deviceIdentification,
+            List<SecurityKeyType> keyTypes) {
+        GetNewSecretsRequest request = new GetNewSecretsRequest();
+        request.setDeviceId(deviceIdentification);
+        request.setSecretTypes(new SecretTypes());
+        List<SecretType> secretTypeList = request.getSecretTypes().getSecretType();
+        keyTypes.stream().forEach(kt -> secretTypeList.add(kt.toSecretType()));
+        return request;
     }
 
+    /*@Override
+    public void aesDecryptAndStoreNewKey(String deviceIdentification, SecurityKeyType keyType, byte[] encryptedKey)
+            throws FunctionalException {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This block of commented-out lines of code should be removed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)